File: src/main/java/net/cjsah/console/command/context/ParsedNodeResult.java
package net.cjsah.console.command.context;

public class ParsedNodeResult<T> {
    private final int start;
    private final int end;
    private final T result;

    public ParsedNodeResult(final int start, final int end, final T result) {
        this.start = start;
        this.end = end;
        this.result = result;
    }

    public IntRange getRange() {
        return new IntRange(this.start, this.end);
    }

    public T getResult() {
        return result;
    }
}

File: src/main/kotlin/net/cjsah/bot/console/command/context/ParsedNodeResult.kt
package net.cjsah.bot.console.command.context

import net.cjsah.bot.console.command.StringReader

class ParsedNodeResult<T>(private val start: Int, private val end: Int, val result: T) {
    fun getNode(reader: StringReader): String {
        return reader.getString().substring(start, end)
    }

    fun getRange(): IntRange {
        return IntRange(start, end)
    }

    fun isEmpty(): Boolean {
        return start == end
    }

    fun getLength(): Int {
        return end - start
    }
}